go/proof_test.go

```diff
@@ -54,3 +54,21 @@ func TestCheckAgainstSpec(t *testing.T) {
 		})
 	}
 }
+
+func TestEmptyBranch(t *testing.T) {
+	cases := EmptyBranchTestData(t)
+
+	for _, tc := range cases {
+		t.Run("case", func(t *testing.T) {
+			if err := tc.Op.CheckAgainstSpec(tc.Spec); err != nil {
+				t.Errorf("Invalid InnerOp: %v", err)
+			}
+			if leftBranchesAreEmpty(tc.Spec.InnerSpec, tc.Op, 0) != tc.IsLeft {
```

Comment:
Hmm, you are calling `leftBranchesAreEmpty` not `IsLeftMost`, so maybe my comments are wrong above.

However if there are no left branch, I would assume that `leftBranchesAreEmpty` would be true

Reply:
I think this is an error of naming and intended meaning. The field should probably have been `IsLeftPlaceholder`, since I meant for this to only be true for ops that contain an empty subtree on the left side.

It didn't make sense to me to call `leftBranchesAreEmpty` on the InnerOp describing a left branch, since by definition its left branch is non empty. So, I mistakenly used the branch index to instead get the padding for the branch on the opposite side (which of course works only when there are 2 children. I made it worse by hard coding the branch to make it work).

Even with your fix though, I would argue that these functions should not pass on Ops that describe a non-empty branch on the corresponding side. They only pass trivially now because the padding is empty, which I'd consider a bug.
So, I propose renaming the field to `IsLeftPlaceholder`, and changing the behavior so the functions fail when there is no padding to check for the corresponding side so e.g. `leftBranchesAreEmpty(..., branch=0)` is always false.

Reply:
Since you will be the largest user if this, I can accept it. I am still not convinced this is the best or most maintainable way to write it.

I would like the function to say "isLeftmost" regardless of empty children to the left and **replace** the current padding check with just this function. Rather than check if it is actually the leftmost, and then check if it is not leftmost, but all (non-zero) left siblings are empty.

I think doing this logic in one piece would avoid confusion later. But I won't block this pr more.